File: snippets/fsharp/System/Type/GetTypeCode/iconvertible.fs
// <snippet1>
open System

/// Class that implements IConvertible
type Complex(x, y) =

    member _.GetDoubleValue() = 
        x ** 2. + y ** 2.
        |> sqrt

    interface IConvertible with
        member _.GetTypeCode() = 
            TypeCode.Object

        member _.ToBoolean(provider: IFormatProvider) =
            (x <> 0.) || (y <> 0.)

        member this.ToByte(provider: IFormatProvider) =
            Convert.ToByte(this.GetDoubleValue())

        member this.ToChar(provider: IFormatProvider) =
            Convert.ToChar(this.GetDoubleValue())

        member this.ToDateTime(provider: IFormatProvider) =
            Convert.ToDateTime(this.GetDoubleValue())

        member this.ToDecimal(provider: IFormatProvider) =
            Convert.ToDecimal(this.GetDoubleValue())

        member this.ToDouble(provider: IFormatProvider) =
            this.GetDoubleValue()

        member this.ToInt16(provider: IFormatProvider) =
            Convert.ToInt16(this.GetDoubleValue())

        member this.ToInt32(provider: IFormatProvider) =
            Convert.ToInt32(this.GetDoubleValue())

        member this.ToInt64(provider: IFormatProvider) =
            Convert.ToInt64(this.GetDoubleValue())

        member this.ToSByte(provider: IFormatProvider) =
            Convert.ToSByte(this.GetDoubleValue())

        member this.ToSingle(provider: IFormatProvider) =
            Convert.ToSingle(this.GetDoubleValue())

        member _.ToString(provider: IFormatProvider) =
            $"( {x} , {y} )"

        member this.ToType(conversionType: Type, provider: IFormatProvider) =
            Convert.ChangeType(this.GetDoubleValue(), conversionType)

        member this.ToUInt16(provider: IFormatProvider) =
            Convert.ToUInt16(this.GetDoubleValue())

        member this.ToUInt32(provider: IFormatProvider) =
            Convert.ToUInt32(this.GetDoubleValue())

        member this.ToUInt64(provider: IFormatProvider) =
            Convert.ToUInt64(this.GetDoubleValue())

// <snippet2>
let writeObjectInfo (testObject: obj) = 
    let typeCode = Type.GetTypeCode(testObject.GetType())
    match typeCode with
    | TypeCode.Boolean ->
        printfn $"Boolean: {testObject}"
    | TypeCode.Double ->
        printfn "Double: {testObject}"
    | _ ->
        printfn $"{typeCode}: {testObject}"
// </snippet2>

let testComplex = Complex(4, 7)

writeObjectInfo testComplex
writeObjectInfo(Convert.ToBoolean(testComplex))
writeObjectInfo(Convert.ToDecimal(testComplex))
writeObjectInfo(Convert.ToString(testComplex))

(*
This code example produces the following results:

Object: ConsoleApplication2.Complex
Boolean: True
Decimal: 8.06225774829855
String: ( 4 , 7 )

*)
// </snippet1>

File: snippets/csharp/System/Type/GetTypeCode/iconvertible.cs
// <snippet1>
using System;

namespace ConsoleApplication2
{

    /// Class that implements IConvertible
    class Complex : IConvertible
    {
        double x;
        double y;

        public Complex(double x, double y)
        {
            this.x = x;
            this.y = y;
        }

        public TypeCode GetTypeCode() => TypeCode.Object;

        bool IConvertible.ToBoolean(IFormatProvider provider)
        {
            if ((x != 0.0) || (y != 0.0))
                return true;
            else
                return false;
        }

        double GetDoubleValue() => Math.Sqrt(x * x + y * y);

        byte IConvertible.ToByte(IFormatProvider provider) => Convert.ToByte(GetDoubleValue());

        char IConvertible.ToChar(IFormatProvider provider) => Convert.ToChar(GetDoubleValue());

        DateTime IConvertible.ToDateTime(IFormatProvider provider) => Convert.ToDateTime(GetDoubleValue());

        decimal IConvertible.ToDecimal(IFormatProvider provider) => Convert.ToDecimal(GetDoubleValue());

        double IConvertible.ToDouble(IFormatProvider provider) => GetDoubleValue();

        short IConvertible.ToInt16(IFormatProvider provider) => Convert.ToInt16(GetDoubleValue());

        int IConvertible.ToInt32(IFormatProvider provider) => Convert.ToInt32(GetDoubleValue());

        long IConvertible.ToInt64(IFormatProvider provider) => Convert.ToInt64(GetDoubleValue());

        sbyte IConvertible.ToSByte(IFormatProvider provider) => Convert.ToSByte(GetDoubleValue());

        float IConvertible.ToSingle(IFormatProvider provider) => Convert.ToSingle(GetDoubleValue());

        string IConvertible.ToString(IFormatProvider provider) => "( " + x.ToString() + " , " + y.ToString() + " )";

        object IConvertible.ToType(Type conversionType, IFormatProvider provider) => Convert.ChangeType(GetDoubleValue(), conversionType);

        ushort IConvertible.ToUInt16(IFormatProvider provider) => Convert.ToUInt16(GetDoubleValue());

        uint IConvertible.ToUInt32(IFormatProvider provider) => Convert.ToUInt32(GetDoubleValue());

        ulong IConvertible.ToUInt64(IFormatProvider provider) => Convert.ToUInt64(GetDoubleValue());
    }

    /// <summary>
    /// Summary description for Class1.
    /// </summary>
    class Class1
    {
        static void Main(string[] args)
        {

            Complex testComplex = new(4, 7);

            WriteObjectInfo(testComplex);
            WriteObjectInfo(Convert.ToBoolean(testComplex));
            WriteObjectInfo(Convert.ToDecimal(testComplex));
            WriteObjectInfo(Convert.ToString(testComplex));
        }
        // <snippet2>
        static void WriteObjectInfo(object testObject)
        {
            TypeCode typeCode = Type.GetTypeCode(testObject.GetType());

            switch (typeCode)
            {
                case TypeCode.Boolean:
                    Console.WriteLine($"Boolean: {testObject}");
                    break;

                case TypeCode.Double:
                    Console.WriteLine($"Double: {testObject}");
                    break;

                default:
                    Console.WriteLine($"{typeCode.ToString()}: {testObject}");
                    break;
            }
        }
        // </snippet2>
    }
}

/*
This code example produces the following results:

Object: ConsoleApplication2.Complex
Boolean: True
Decimal: 8.06225774829855
String: ( 4 , 7 )

*/
// </snippet1>
